### PR TITLE
feat/LW-7903: Add Pool Id to EpochRewards interface

### PR DESCRIPTION
**Pull request by AngelCastilloB**
# Context

Now that we have multi delegation, we want to have the ability to segregate epoch rewards by pools

# Proposed Solution

The rewards table already has the pool id of the pool that generated the rewards:

### `reward`

A table for earned rewards. It includes 5 types of rewards. The rewards are inserted incrementally and this procedure is finalised when the spendable epoch comes. Before the epoch comes, some entries may be missing.

* Primary Id: `id`

| Column name | Type | Description |
|-|-|-|
| `id` | integer (64) |  |
| `addr_id` | integer (64) | The StakeAddress table index for the stake address that earned the reward. |
| `type` | rewardtype | The source of the rewards; pool `member`, pool `leader`, `treasury` or `reserves` payment and pool deposits `refunds` |
| `amount` | lovelace | The reward amount (in Lovelace). |
| `earned_epoch` | integer (64) | The epoch in which the reward was earned. For `pool` and `leader` rewards spendable in epoch `N`, this will be `N - 2`, for `treasury` and `reserves` `N - 1` and for `refund` N. |
| `spendable_epoch` | integer (64) | The epoch in which the reward is actually distributed and can be spent. |
| **`pool_id`** | integer (64) | The PoolHash table index for the pool the stake address was delegated to when the reward is earned or for the pool that there is a deposit refund. Will be NULL for payments from the treasury or the reserves. |

All we need to do is join the pool hash table to get the bech32 pool id and include it in the result:

### `pool_hash`

A table for every unique pool key hash. The `id` field of this table is used as foreign keys in other tables. The existance of an entry doesn't mean the pool is registered or in fact that is was ever registered.

* Primary Id: `id`

| Column name | Type | Description |
|-|-|-|
| `id` | integer (64) |  |
| `hash_raw` | hash28type | The raw bytes of the pool hash. |
| **`view`** | string | The Bech32 encoding of the pool hash. |

For example:

```
  WITH epochs AS (
      SELECT 
          id,
          no AS epoch_no
      FROM epoch
  )
  SELECT
      SUM(r.amount) AS quantity,
      sa."view" AS address,
      r.earned_epoch AS epoch,
      ph."view" as poolId
  FROM reward r
  JOIN stake_address sa ON
      sa.id = r.addr_id 
  JOIN epochs ON 
      r.earned_epoch = epochs.epoch_no
  JOIN pool_hash ph ON 
      r.pool_id = ph.id
  GROUP BY sa."view", r.earned_epoch, ph."view"
  ORDER BY quantity DESC
  limit 5
```

|`quantity`	|`address`	|`epoch`	|`poolid`
|-|-|-|-|
10597898673019|	"stake_test1uzrq54n7feexz8znrtz5w0patfpuuc4av563e8jzqzxgzucuzzlur"	|1	|"pool1h8yl5mkyrfmfls2x9fu9mls3ry6egnw4q6efg34xr37zc243gkf"
9991108215335|	"stake_test1uz0fqdpdu38t907z6erurdn4ht8phzxtrv3jweccthwmdvs3htnvz"	|4	|"pool1xauwunq374detlph3ujw8ncnumcxu9vr475l3yf0nhskgheq6d8"
9939507263871	|"stake_test1urdz0aunucthftlftzcgrc3wah5tnhhkmk9s92hcxzfkg5cvmd8zl"	|5	|"pool1r3c4nyskw28ethhjxrc89fdwlcqw4tf076x7q9vd5c3h5qrgu4n"
9560081730187|	"stake_test1uzrq54n7feexz8znrtz5w0patfpuuc4av563e8jzqzxgzucuzzlur"	|2	|"pool1h8yl5mkyrfmfls2x9fu9mls3ry6egnw4q6efg34xr37zc243gkf"
9560080172248	|"stake_test1uqkzl6vmywlvufknuaedfnkm2h2zeacv38wrv9y7s2y330geclc03"|	2	|"pool1n7xfve4q5m3j9300nadwwzarxumsakhge7gkegcyhr8qjwa3vz9"

# Important Changes Introduced

- EpochRewards renamed to Reward
- Reward now includes the poolId of the pool the stake address was delegated to when the reward is earned. (Will be undefined for payments from the treasury or the reserves.)
- Reward no longer coalesce rewards from the same epoch

### CODE REVIEW

File: packages/core/src/Provider/RewardsProvider/types.ts [hunk 1]
```diff
@@ -5,6 +5,11 @@ import { Range } from '@cardano-sdk/util';
 export interface EpochRewards {
   epoch: Cardano.EpochNo;
   rewards: Cardano.Lovelace;
+  /**
+   * The pool the stake address was delegated to when the reward is earned.
+   * Will be null for payments from the treasury or the reserves.
+   */
+  poolId: Cardano.PoolId | null;
```
Comment (mkazlauskas) [inline packages/core/src/Provider/RewardsProvider/types.ts:12]: This makes it forward-compatible (techically), as old providers would return objects with this `undefined`
```suggestion
  poolId?: Cardano.PoolId;
```

File: packages/util-dev/src/mockProviders/mockData.ts [hunk 2]
```diff
@@ -42,14 +42,16 @@ export const protocolParameters = {
 export const epochRewards = [
   {
     epoch: Cardano.EpochNo(currentEpoch.number - 3),
+    poolId: Cardano.PoolId('pool1h8yl5mkyrfmfls2x9fu9mls3ry6egnw4q6efg34xr37zc243gkf'),
```
Comment (iccicci) [inline packages/util-dev/src/mockProviders/mockData.ts:45]: I see this pool id more times in this PR.
Is this mock the only source of this pool id or is it taken from fixtures (db dump), too?
I'm asking just to be sure this PR is resilient to fixtures re-generation.
Thank you

**Review by mkazlauskas — CHANGES_REQUESTED** [review]
When squashing, please update commit message footer to the following format:
```
BREAKING CHANGE: change1
- change2
- change3
```
Then it's picked up to be included in the CHANGELOG and formatted well

File: packages/cardano-services/src/Rewards/DbSyncRewardProvider/queries.ts [hunk 3]
```diff
@@ -40,14 +40,17 @@ export const findRewardsHistory = (lowerBound?: number, upperBound?: number) =>
...
   FROM reward r
   JOIN stake_address sa ON
       sa.id = r.addr_id AND 
       sa."view" = ANY($1)
   JOIN epochs ON 
       r.earned_epoch = epochs.epoch_no
-  GROUP BY sa."view", r.earned_epoch
+  JOIN pool_hash ph ON 
+      r.pool_id = ph.id
+  GROUP BY sa."view", r.earned_epoch, ph."view"
```
[the remark this answers is not in the document]
Comment (iccicci) [reply packages/cardano-services/src/Rewards/DbSyncRewardProvider/queries.ts:53]: Apart from the logic (which, from my point of view, seems to be correct as well),
each filed in the `SELECT` list must be present in the `GROUP BY` list or aggregated with an aggregation function (`SUM`, `AVG`, `MIN`, `MAX`, etc.)

[the remark this answers is not in the document]
Comment (mkazlauskas) [reply packages/cardano-services/src/Rewards/DbSyncRewardProvider/queries.ts:53]: Got it, thanks guys

**Review by mkazlauskas — APPROVED** [review]
Great work! :rocket:

**Review by iccicci — APPROVED** [review]
LGTM!

**Pull request merged**